src/trio/_path.py

```diff
@@ -32,219 +22,102 @@
         OpenBinaryModeWriting,
         OpenTextMode,
     )
-    from typing_extensions import Concatenate, Literal, ParamSpec, TypeAlias
-
-    from trio._file_io import AsyncIOWrapper as _AsyncIOWrapper
+    from typing_extensions import Concatenate, Literal, ParamSpec
 
     P = ParamSpec("P")
 
-T = TypeVar("T")
-StrPath: TypeAlias = Union[str, "os.PathLike[str]"]  # Only subscriptable in 3.9+
-
+    PathT = TypeVar("PathT", bound="Path")
+    T = TypeVar("T")
 
-# re-wrap return value from methods that return new instances of pathlib.Path
-def rewrap_path(value: T) -> T | Path:
-    if isinstance(value, pathlib.Path):
-        return Path(value)
-    else:
-        return value
 
+def _wraps_async(
+    wrapped: Callable[..., Any]
+) -> Callable[[Callable[P, T]], Callable[P, Awaitable[T]]]:
+    def decorator(fn: Callable[P, T]) -> Callable[P, Awaitable[T]]:
+        async def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
+            return await run_sync(partial(fn, *args, **kwargs))
 
-def _forward_factory(
-    cls: AsyncAutoWrapperType,
-    attr_name: str,
-    attr: Callable[Concatenate[pathlib.Path, P], T],
-) -> Callable[Concatenate[Path, P], T | Path]:
-    @wraps(attr)
-    def wrapper(self: Path, *args: P.args, **kwargs: P.kwargs) -> T | Path:
-        attr = getattr(self._wrapped, attr_name)
-        value = attr(*args, **kwargs)
-        return rewrap_path(value)
+        update_wrapper(wrapper, wrapped)
+        assert wrapped.__doc__ is not None
+        wrapper.__doc__ = (
+            f"Like :meth:`~{wrapped.__module__}.{wrapped.__qualname__}`, but async.\n"
+            f"\n"
+            f"{cleandoc(wrapped.__doc__)}\n"
+        )
+        return wrapper
 
-    # Assigning this makes inspect and therefore Sphinx show the original parameters.
-    # It's not defined on functions normally though, this is a custom attribute.
-    assert isinstance(wrapper, types.FunctionType)
-    wrapper.__signature__ = inspect.signature(attr)
-
-    return wrapper
+    return decorator
 
 
-def _forward_magic(
-    cls: AsyncAutoWrapperType, attr: Callable[..., T]
-) -> Callable[..., Path | T]:
-    sentinel = object()
+def _wrap_method(
+    fn: Callable[Concatenate[pathlib.Path, P], T],
+) -> Callable[Concatenate[Path, P], Awaitable[T]]:
+    @_wraps_async(fn)
+    def wrapper(self: Path, /, *args: P.args, **kwargs: P.kwargs) -> T:
+        return fn(pathlib.Path(self), *args, **kwargs)
```

Comment:
Doing `pathlib.Path(self)` rather than caching it as `self._wrapped` might have implications? I'm not familiar with what the `pathlib.Path` `__init__` does though.

Reply:
`Path` is spammed all over the place, as it's an immutable class and the everyday APIs create new ones all the time. Something like: `foo / bar / baz` creates a new `Path` and then throws it away at each intermediate `/` time.

So it's best to optimize this for cheap creation of new `trio.Path` objects, which is what this PR does. The new `trio.Path` is slotted, and does not wrap another class, so I'd expect non-blocking operations on the new `trio.Path` to be quite a bit faster.

The actual blocking operations are gonna be slow anyway. Compare to creating a `partial` and passing it to a thread via a thread-safe channel and then waiting for the result using an event loop, I'd expect it to be trivial.

If you'd like me to run some benchmarks, I'd be happy to. But my instinct here is that non-blocking operations will be faster, and wrapped blocking operations will functionally identical.

Reply:
No benchmarks necessary. I was more thinking about any possible exceptions being raised!